File: frontend/lib/group-entities-related-parties/rpt.ts
/**
 * RPT calculation helpers — frontend-only, decimal-safe.
 */

import { addDecimals, isFilledDecimal, parseDecimal, percentageOf } from '@/lib/group-entities-related-parties/decimal';
import type { LinkedWorkstreamReferences } from '@/lib/group-entities-related-parties/types';
import type {
  GroupEntitiesRelatedPartiesPayload,
  RptBalanceRecord,
  RptTransactionRecord,
} from '@/lib/schemas/group-entities-related-parties';

const SALES_TYPES = new Set([
  'sale-of-goods-materials',
  'sale-of-services',
  'property-sale',
  'lease-rent-received',
  'licence-royalty',
  'dividend',
]);

const PURCHASE_TYPES = new Set([
  'purchase-of-goods-materials',
  'purchase-receipt-of-services',
  'property-purchase',
  'lease-rent-paid',
  'management-services',
]);

const LOAN_GIVEN_TYPES = new Set(['loan-given', 'advance-given']);
const LOAN_RECEIVED_TYPES = new Set(['loan-received', 'advance-received']);
const GUARANTEE_TYPES = new Set(['guarantee', 'corporate-guarantee', 'collateral-security']);

export type RptSummary = {
  totalByParty: Record<string, string>;
  totalByType: Record<string, string>;
  totalByFinancialYear: Record<string, string>;
  rptSales: string;
  rptPurchases: string;
  rptLoansGiven: string;
  rptLoansReceived: string;
  guarantees: string;
  closingReceivables: string;
  closingPayables: string;
  closingLoans: string;
  latestFinancialYearTotal: string;
  rptRevenuePercent: string | null;
  rptPurchasesPercent: string | null;
  rptReceivablesPercent: string | null;
  rptPayablesPercent: string | null;
  financialsRevenueDifference: string | null;
  financialsPurchasesDifference: string | null;
};

function partyKey(tx: RptTransactionRecord): string {
  return tx.linkedEntityId || tx.linkedPersonId || tx.relatedPartyRelationshipId || 'unknown';
}

function sumTransactions(
  transactions: RptTransactionRecord[],
  filter: (tx: RptTransactionRecord) => boolean,
): string {
  const values = transactions.filter(filter).map((tx) => tx.transactionValue);
  return addDecimals(...values);
}

function sumBalances(balances: RptBalanceRecord[], types: Set<string>): string {
  const values = balances
    .filter((balance) => types.has(balance.balanceType))
    .map((balance) => balance.closingBalance);
  return addDecimals(...values);
}

export function calculateRptSummary(
  payload: GroupEntitiesRelatedPartiesPayload,
  linkedReferences: LinkedWorkstreamReferences,
): RptSummary {
  const transactions = payload.relatedPartyTransactionsBalancesAndCommitments.transactions;
  const balances = payload.relatedPartyTransactionsBalancesAndCommitments.balances;

  const totalByParty: Record<string, string> = {};
  const totalByType: Record<string, string> = {};
  const totalByFinancialYear: Record<string, string> = {};

  for (const tx of transactions) {
    const key = partyKey(tx);
    totalByParty[key] = addDecimals(totalByParty[key] ?? '', tx.transactionValue);
    if (tx.transactionType) {
      totalByType[tx.transactionType] = addDecimals(
        totalByType[tx.transactionType] ?? '',
        tx.transactionValue,
      );
    }
    if (tx.financialPeriod) {
      totalByFinancialYear[tx.financialPeriod] = addDecimals(
        totalByFinancialYear[tx.financialPeriod] ?? '',
        tx.transactionValue,
      );
    }
  }

  const rptSales = sumTransactions(transactions, (tx) => SALES_TYPES.has(tx.transactionType));
  const rptPurchases = sumTransactions(transactions, (tx) => PURCHASE_TYPES.has(tx.transactionType));
  const rptLoansGiven = sumTransactions(transactions, (tx) => LOAN_GIVEN_TYPES.has(tx.transactionType));
  const rptLoansReceived = sumTransactions(transactions, (tx) =>
    LOAN_RECEIVED_TYPES.has(tx.transactionType),
  );
  const guarantees = sumTransactions(transactions, (tx) => GUARANTEE_TYPES.has(tx.transactionType));

  const closingReceivables = sumBalances(
    balances,
    new Set(['receivable', 'loan-receivable', 'advance', 'accrued-income']),
  );
  const closingPayables = sumBalances(
    balances,
    new Set(['payable', 'loan-payable', 'accrued-expense', 'commitment']),
  );
  const closingLoans = addDecimals(
    sumBalances(balances, new Set(['loan-receivable'])),
    sumBalances(balances, new Set(['loan-payable'])),
  );

  const financialYears = Object.keys(totalByFinancialYear).sort();
  const latestFinancialYear = financialYears.at(-1) ?? '';
  const latestFinancialYearTotal = latestFinancialYear
    ? totalByFinancialYear[latestFinancialYear]
    : '';

  const fin = linkedReferences.financialsKpis;
  const revenueFromOperations = fin.revenueFromOperations ?? '';
  const totalPurchases = fin.totalPurchases ?? '';
  const totalReceivables = fin.totalReceivables ?? '';
  const totalPayables = fin.totalPayables ?? '';
  const rptRevenueTotal = fin.rptRevenueTotal ?? '';
  const rptPurchasesTotal = fin.rptPurchasesTotal ?? '';

  const rptRevenuePercent =
    fin.available && isFilledDecimal(revenueFromOperations)
      ? percentageOf(rptSales, revenueFromOperations)
      : null;
  const rptPurchasesPercent =
    fin.available && isFilledDecimal(totalPurchases)
      ? percentageOf(rptPurchases, totalPurchases)
      : null;
  const rptReceivablesPercent =
    fin.available && isFilledDecimal(totalReceivables)
      ? percentageOf(closingReceivables, totalReceivables)
      : null;
  const rptPayablesPercent =
    fin.available && isFilledDecimal(totalPayables)
      ? percentageOf(closingPayables, totalPayables)
      : null;

  let financialsRevenueDifference: string | null = null;
  if (fin.available && isFilledDecimal(rptRevenueTotal) && isFilledDecimal(rptSales)) {
    const finVal = parseDecimal(rptRevenueTotal)!;
    const calcVal = parseDecimal(rptSales)!;
    financialsRevenueDifference = String(Math.abs(finVal - calcVal));
  }

  let financialsPurchasesDifference: string | null = null;
  if (fin.available && isFilledDecimal(rptPurchasesTotal) && isFilledDecimal(rptPurchases)) {
    const finVal = parseDecimal(rptPurchasesTotal)!;
    const calcVal = parseDecimal(rptPurchases)!;
    financialsPurchasesDifference = String(Math.abs(finVal - calcVal));
  }

  return {
    totalByParty,
    totalByType,
    totalByFinancialYear,
    rptSales,
    rptPurchases,
    rptLoansGiven,
    rptLoansReceived,
    guarantees,
    closingReceivables,
    closingPayables,
    closingLoans,
    latestFinancialYearTotal,
    rptRevenuePercent,
    rptPurchasesPercent,
    rptReceivablesPercent,
    rptPayablesPercent,
    financialsRevenueDifference,
    financialsPurchasesDifference,
  };
}

File: backend/app/modules/group_entities_related_parties/rpt.py
"""RPT calculation helpers — decimal-safe."""

from __future__ import annotations

from typing import Any

from app.modules.group_entities_related_parties import decimal_utils as dm

SALES_TYPES = frozenset(
    {
        "sale-of-goods-materials",
        "sale-of-services",
        "property-sale",
        "lease-rent-received",
        "licence-royalty",
        "dividend",
    }
)

PURCHASE_TYPES = frozenset(
    {
        "purchase-of-goods-materials",
        "purchase-receipt-of-services",
        "property-purchase",
        "lease-rent-paid",
        "management-services",
    }
)

LOAN_GIVEN_TYPES = frozenset({"loan-given", "advance-given"})
LOAN_RECEIVED_TYPES = frozenset({"loan-received", "advance-received"})
GUARANTEE_TYPES = frozenset({"guarantee", "corporate-guarantee", "collateral-security"})

RECEIVABLE_BALANCE_TYPES = frozenset(
    {"receivable", "loan-receivable", "advance", "accrued-income"}
)
PAYABLE_BALANCE_TYPES = frozenset(
    {"payable", "loan-payable", "accrued-expense", "commitment"}
)


def _party_key(tx: dict[str, Any]) -> str:
    return (
        str(tx.get("linkedEntityId") or "")
        or str(tx.get("linkedPersonId") or "")
        or str(tx.get("relatedPartyRelationshipId") or "")
        or "unknown"
    )


def _sum_transactions(
    transactions: list[dict[str, Any]],
    predicate,
) -> str:
    values = [
        str(tx.get("transactionValue") or "")
        for tx in transactions
        if isinstance(tx, dict) and predicate(tx)
    ]
    return dm.sum_decimals(values)


def _sum_balances(balances: list[dict[str, Any]], types: frozenset[str]) -> str:
    values = [
        str(balance.get("closingBalance") or "")
        for balance in balances
        if isinstance(balance, dict) and balance.get("balanceType") in types
    ]
    return dm.sum_decimals(values)


def _abs_difference(a: str, b: str) -> str | None:
    parsed_a = dm.parse_decimal(a)
    parsed_b = dm.parse_decimal(b)
    if parsed_a is None or parsed_b is None:
        return None
    return str(abs(parsed_a - parsed_b))


def calculate_rpt_summary(
    payload: dict[str, Any],
    linked_references: dict[str, Any],
) -> dict[str, Any]:
    rpt_section = payload.get("relatedPartyTransactionsBalancesAndCommitments") or {}
    transactions = [tx for tx in (rpt_section.get("transactions") or []) if isinstance(tx, dict)]
    balances = [b for b in (rpt_section.get("balances") or []) if isinstance(b, dict)]

    total_by_party: dict[str, str] = {}
    total_by_type: dict[str, str] = {}
    total_by_financial_year: dict[str, str] = {}

    for tx in transactions:
        key = _party_key(tx)
        total_by_party[key] = dm.sum_decimals([total_by_party.get(key, ""), str(tx.get("transactionValue") or "")])
        tx_type = str(tx.get("transactionType") or "")
        if tx_type:
            total_by_type[tx_type] = dm.sum_decimals(
                [total_by_type.get(tx_type, ""), str(tx.get("transactionValue") or "")]
            )
        financial_period = str(tx.get("financialPeriod") or "")
        if financial_period:
            total_by_financial_year[financial_period] = dm.sum_decimals(
                [
                    total_by_financial_year.get(financial_period, ""),
                    str(tx.get("transactionValue") or ""),
                ]
            )

    rpt_sales = _sum_transactions(transactions, lambda tx: tx.get("transactionType") in SALES_TYPES)
    rpt_purchases = _sum_transactions(
        transactions,
        lambda tx: tx.get("transactionType") in PURCHASE_TYPES,
    )
    rpt_loans_given = _sum_transactions(
        transactions,
        lambda tx: tx.get("transactionType") in LOAN_GIVEN_TYPES,
    )
    rpt_loans_received = _sum_transactions(
        transactions,
        lambda tx: tx.get("transactionType") in LOAN_RECEIVED_TYPES,
    )
    guarantees = _sum_transactions(
        transactions,
        lambda tx: tx.get("transactionType") in GUARANTEE_TYPES,
    )

    closing_receivables = _sum_balances(balances, RECEIVABLE_BALANCE_TYPES)
    closing_payables = _sum_balances(balances, PAYABLE_BALANCE_TYPES)
    closing_loans = dm.sum_decimals(
        [
            _sum_balances(balances, frozenset({"loan-receivable"})),
            _sum_balances(balances, frozenset({"loan-payable"})),
        ]
    )

    financial_years = sorted(total_by_financial_year.keys())
    latest_financial_year = financial_years[-1] if financial_years else ""
    latest_financial_year_total = (
        total_by_financial_year.get(latest_financial_year, "") if latest_financial_year else ""
    )

    fin = linked_references.get("financialsKpis") or {}
    revenue_from_operations = str(fin.get("revenueFromOperations") or "")
    total_purchases = str(fin.get("totalPurchases") or "")
    total_receivables = str(fin.get("totalReceivables") or "")
    total_payables = str(fin.get("totalPayables") or "")
    rpt_revenue_total = str(fin.get("rptRevenueTotal") or "")
    rpt_purchases_total = str(fin.get("rptPurchasesTotal") or "")

    fin_available = bool(fin.get("available"))

    rpt_revenue_percent = (
        dm.percentage_of(rpt_sales, revenue_from_operations)
        if fin_available and dm.is_filled(revenue_from_operations)
        else None
    )
    rpt_purchases_percent = (
        dm.percentage_of(rpt_purchases, total_purchases)
        if fin_available and dm.is_filled(total_purchases)
        else None
    )
    rpt_receivables_percent = (
        dm.percentage_of(closing_receivables, total_receivables)
        if fin_available and dm.is_filled(total_receivables)
        else None
    )
    rpt_payables_percent = (
        dm.percentage_of(closing_payables, total_payables)
        if fin_available and dm.is_filled(total_payables)
        else None
    )

    financials_revenue_difference = None
    if fin_available and dm.is_filled(rpt_revenue_total) and dm.is_filled(rpt_sales):
        financials_revenue_difference = _abs_difference(rpt_revenue_total, rpt_sales)

    financials_purchases_difference = None
    if fin_available and dm.is_filled(rpt_purchases_total) and dm.is_filled(rpt_purchases):
        financials_purchases_difference = _abs_difference(rpt_purchases_total, rpt_purchases)

    return {
        "totalByParty": total_by_party,
        "totalByType": total_by_type,
        "totalByFinancialYear": total_by_financial_year,
        "rptSales": rpt_sales,
        "rptPurchases": rpt_purchases,
        "rptLoansGiven": rpt_loans_given,
        "rptLoansReceived": rpt_loans_received,
        "guarantees": guarantees,
        "closingReceivables": closing_receivables,
        "closingPayables": closing_payables,
        "closingLoans": closing_loans,
        "latestFinancialYearTotal": latest_financial_year_total,
        "rptRevenuePercent": rpt_revenue_percent,
        "rptPurchasesPercent": rpt_purchases_percent,
        "rptReceivablesPercent": rpt_receivables_percent,
        "rptPayablesPercent": rpt_payables_percent,
        "financialsRevenueDifference": financials_revenue_difference,
        "financialsPurchasesDifference": financials_purchases_difference,
    }
